streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java

```diff
@@ -0,0 +1,230 @@
+            throw new IllegalArgumentException("storeName cannot be null");
+        }
+        final Set<String> sourceTopics = builder.stateStoreNameToSourceTopics().get(storeName);
+        if (sourceTopics == null) {
+            return Collections.emptyList();
+        }
+
+        final ArrayList<StreamsMetadata> results = new ArrayList<>();
+        for (StreamsMetadata instance : allMetadata) {
+            if (instance.stateStoreNames().contains(storeName)) {
+                results.add(instance);
+            }
+        }
+        return results;
+    }
+
+    /**
+     * Find the {@link StreamsMetadata}s for a given storeName and key.
+     * Note: the key may not exist in the {@link org.apache.kafka.streams.processor.StateStore},
+     * this method provides a way of finding which {@link StreamsMetadata} it would exist on.
+     *
+     * @param storeName     Name of the store
+     * @param key           Key to use to for partition
+     * @param keySerializer Serializer for the key
+     * @param <K>           key type
+     * @return The {@link StreamsMetadata} for the storeName and key
+     */
+    public synchronized <K> StreamsMetadata getMetadataWithKey(final String storeName,
+                                                               final K key,
+                                                               final Serializer<K> keySerializer) {
+        if (keySerializer == null) {
+            throw new IllegalArgumentException("keySerializer cannot be null");
+        }
+
+        // To find the correct instance for a given key we need to use the same partitioning
+        // strategy that was used when producing the messages. In this case we create a StreamPartitioner
+        // that works the same as org.apache.kafka.clients.producer.internals.DefaultPartitioner
+        return getMetadataWithKey(storeName, key, new StreamPartitioner<K, Object>() {
+            @Override
+            public Integer partition(final K key, final Object value, final int numPartitions) {
+                final String sourceTopic = builder.stateStoreNameToSourceTopics().get(storeName).iterator().next();
+                final byte[] bytes = keySerializer.serialize(sourceTopic, key);
+                return toPositive(Utils.murmur2(bytes)) % numPartitions;
+            }
+        });
+    }
+
+    /**
+     * Find the {@link StreamsMetadata}s for a given storeName and key.
+     * Note: the key may not exist in the {@link StateStore},
+     * this method provides a way of finding which {@link StreamsMetadata} it would exist on.
+     *
+     * @param storeName   Name of the store
+     * @param key         Key to use to for partition
+     * @param partitioner partitioner to use to find correct partition for key
+     * @param <K>         key type
+     * @return The {@link StreamsMetadata} for the storeName and key
+     */
+    public synchronized <K> StreamsMetadata getMetadataWithKey(final String storeName,
```

Comment:
Correct, that is why we let them pass in a partitioner here. There is not a lot we can do about it as we don't have control. But they developer would need to know they used a different partitioner and then they can use the correct one to find the metadata for key.
So it is safe if you either always used `DefaultPartitioner` or you query with the custom partitioner that was used to produce to the topics.